### PR TITLE
Connect-DbaInstance / new code path - General improvements

### LINKED ISSUE #7425: Connect-DbaInstance - Be able to use server SMO and change ApplicationIntent or StatementTimeout
As with NonPooledConnection or Database, when I start Connect-DbaInstance with an already connected server SMO as SqlInstance, then I would like to get a new server SMO back, that uses my given ApplicationIntent or StatementTimeout.

Should be implemented the same way as with Database or NonPooledConnection, so the resulting code should look like this:
```
# Test if we have to copy the connection context
# Currently only if we have a different Database or have to swith to a NonPooledConnection or using a specific StatementTimeout or using ApplicationIntent
# We do not test for SqlCredential as this would change the behavior compared to the legacy code path
$copyContext = $false
if ($Database -and $inputObject.ConnectionContext.CurrentDatabase -ne $Database) {
	Write-Message -Level Verbose -Message "Parameter Database passed in, and it's not the same as currently in ConnectionContext.CurrentDatabase, so we copy the connection context"
	$copyContext = $true
}
if ($ApplicationIntent) {
	Write-Message -Level Verbose -Message "Parameter ApplicationIntent passed in, so we copy the connection context and set the ApplicationIntent"
	$copyContext = $true
}
if ($NonPooledConnection -and -not $inputObject.ConnectionContext.NonPooledConnection) {
	Write-Message -Level Verbose -Message "Parameter NonPooledConnection passed in and we currently have a pooled connection, so we copy the connection context and set NonPooledConnection"
	$copyContext = $true
}
if (Test-Bound -Parameter StatementTimeout) {
	Write-Message -Level Verbose -Message "Parameter StatementTimeout passed in, so we copy the connection context and set the StatementTimeout"
	$copyContext = $true
}
if ($copyContext) {
	$connContext = $inputObject.ConnectionContext.Copy()
	if ($Database) {
		$connContext = $connContext.GetDatabaseConnection($Database)
	}
	if ($ApplicationIntent) {
		$connContext.ApplicationIntent = $ApplicationIntent
	}
	if ($NonPooledConnection) {
		$connContext.NonPooledConnection = $true
	}
	if (Test-Bound -Parameter StatementTimeout) {
		$connContext.StatementTimeout = $StatementTimeout
	}
	$server = New-Object -TypeName Microsoft.SqlServer.Management.Smo.Server -ArgumentList $connContext
} else {
	$server = $inputObject
}
```

The paramater ApplicationIntent is used in Invoke-DbaQuery.

The parameter StatementTimeout is used in these commands:
- Backup-DbaDatabase
- Restore-DbaDatabase
- Get-DbaTopResourceUsage
- Import-DbaCsv
- Invoke-DbaDbLogShipping
- Invoke-DbaDbShrink
- Invoke-DbaDbUpgrade
- Set-DbaDbCompression
- Test-DbaDbCompression
- Start-DbccCheck

There it is mostly used not as a parameter to the typical `$server = Connect-DbaInstance -SqlInstance $instance ...`, but by later changing the server SMO in $server with `$Server.ConnectionContext.StatementTimeout = 0`. This should be changed later for all of those commands.

### LINKED ISSUE #7426: Connect-DbaInstance - Be able to decide which source property should be used as the property ComputerName
I would like to be more flexible choose the suitable ComputerName for my environment.

Current rule: Use $server.NetName, but if $server.NetName is empty or we are on Azure or Linux, use $instance.ComputerName

Examples:
If I'm in a fairly normal environment, the default is fine. But when I'm only working with Azure databases, maybe I would like to not use the long `$instance.ComputerName` "MyAzureSQLServer.database.windows.net", but only the short `$server.DomainInstanceName` "MyAzureSQLServer" for the typical first output property of dbatools commands. And maybe I have a complex docker environment where the `$server.NetName` gives a readable output. And if it is a complex environment with multiple servers with the same name but different domains I maybe want to use the full name in `$instance.ComputerName`.

To achieve this, I propose the following code:

```
if (-not $server.ComputerName) {
	# To set the source of ComputerName to something else than the default use this config parameter:
	# Set-DbatoolsConfig -FullName commands.connect-dbainstance.smo.computername.source -Value 'server.ComputerNamePhysicalNetBIOS'
	# Set-DbatoolsConfig -FullName commands.connect-dbainstance.smo.computername.source -Value 'instance.ComputerName'
	# If the config parameter is not used, then there a different ways to handle the new property ComputerName
	# Rules in legacy code: Use $server.NetName, but if $server.NetName is empty or we are on Azure or Linux, use $instance.ComputerName
	$computerName = $null
	$computerNameSource = Get-DbatoolsConfigValue -FullName commands.connect-dbainstance.smo.computername.source
	if ($computerNameSource) {
		Write-Message -Level Debug -Message "Setting ComputerName based on $computerNameSource"
		$object, $property = $computerNameSource -split '\.'
		$value = (Get-Variable -Name $object).Value.$property
		if ($value) {
			$computerName = $value
			Write-Message -Level Debug -Message "ComputerName will be set to $computerName"
		} else {
			Write-Message -Level Debug -Message "No value found for ComputerName, so will use the default"
		}
	}
	if (-not $computerName) {
		if ($server.DatabaseEngineType -eq "SqlAzureDatabase") {
			Write-Message -Level Debug -Message "We are on Azure, so server.ComputerName will be set to instance.ComputerName"
			$computerName = $instance.ComputerName
		} elseif ($server.HostPlatform -eq 'Linux') {
			Write-Message -Level Debug -Message "We are on Linux what is often on docker and the internal name is not useful, so server.ComputerName will be set to instance.ComputerName"
			$computerName = $instance.ComputerName
		} else {
			Write-Message -Level Debug -Message "We will set server.ComputerName to server.NetName"
			$computerName = $server.NetName
		}
		Write-Message -Level Debug -Message "ComputerName will be set to $computerName"
	}
	Add-Member -InputObject $server -NotePropertyName ComputerName -NotePropertyValue $computerName -Force
}
```

### PR DESCRIPTION
**Pull request by andreasjordan**
## Type of Change
 - [ ] Bug fix (non-breaking change, fixes # )
 - [x] New feature (non-breaking change, adds functionality, fixes #7425, fixes #7426 )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 
### Purpose
All necessary changes to later use the new code path as the default.

### Approach
I addressed most of the TODO entries and added some new features, so beeing able to use more parameters together with a server SMO as SqlInstance.

I would like to discuss the different topics in the linked issues to have a better overview.
If the majority of the maintainers rejects the changes related to the linked issues, I will of course remove them.

### CODE REVIEW

**Comment by potatoqualitee** [comment]
thanks much. Is this the same code that was in experimental for a while? if so, i can merge it then hold off on publishing dev for a few days

**Comment by andreasjordan** [comment]
No, new functionality has also been added. I'll walk through the diff, naming the new lines.
* 476 - 537: Only documentation
* 581 - 632: New functionality, see #7425
* 639 - 658: Additional tests for parameters that are not used anymore by the new code path.
* 661 - 662: No support for MFA inside of the command, possible with outside usage of the AZ module and AccessToken
* 698 - 899: Changed or added some debug messages
* 907 - 919: I changed the way the connection is opened. After many tests this seems to be we best way.
* 940 - 972: New functionality, but also changing the default behavior of the new code path to align it with the old path. But for those how have used the new code path there maybe situations where they get a different value for ComputerName. But with the new config parameter now every desired value can be achieved. See #7426 
* 974 - 1007: New functionality, see #7427 
* 1028 - 1038: Just copied from the old code path and added some documentation
* 1040 - 1075: Copied from the old code path but hiding it behind a configuration. I can't see an effect of this, have traced the sql server and compared what calls are send to the server on connect - I don't see any difference. Because I don't like code where I don't know that it has a positive effect, but on the other hand I just might not see it - I choose this. As soon as someone can prove and show an example where this has a positive effect - we just enable it by default. If in a one year time, nobody came to us - we delete the code and copy it to an issue on github for backup.

So wrapping up: I would like to have your opinion on #7426 and #7427 and of cource the other functional changes. If we have the code we want, then I would like to have these changes for a month or two in the non-default experimental code path. But I don't have any further changes, so the next step then would be just to change the default and add a config parameter to activate the legacy code path as a backup.

**Comment by andreasjordan** [comment]
If you would like to talk about the changes, I can set up a Teams meeting and we can go through the code together.

**Comment by potatoqualitee** [comment]
This is incredible, thank you so much. Yes, I'll likely need a teams meeting @andreasjordan 💯  Perhaps late next week? This is intense and I can't think in this heat.

**Comment by andreasjordan** [comment]
Just ping me on slack if you have time.

File: functions/Connect-DbaInstance.ps1 [hunk 1]
```diff
@@ -462,36 +462,58 @@ function Connect-DbaInstance {
             and within this function leads to the following code path being used.
             All the sub paths inside the following if clause will end with a continue, so the normal code path is not used.
             #>
+
+            # The following line is just added so that we can use AppVayor to run all test with the experimental code path
+            Set-DbatoolsConfig -FullName sql.connection.experimental -Value $true
```
Comment (wsmelton) [inline functions/Connect-DbaInstance.ps1:467]: FYI, if you want Appveyor to run using this easier to put it in the Test itself.

File: functions/Connect-DbaInstance.ps1 [hunk 2]
```diff
@@ -665,26 +725,21 @@ function Connect-DbaInstance {
...
-                    if ($authType -eq 'AD Universal with MFA Support') {
-                        # Azure AD with Multi-Factor Authentication
-                        # TODO: This is not tested
-                        Write-Message -Level Debug -Message "Authentication will be set to 'ActiveDirectoryInteractive'"
-                        $connInfo.Authentication = [Microsoft.SqlServer.Management.Common.SqlConnectionInfo+AuthenticationMethod]::ActiveDirectoryInteractive
-                    } elseif ($authType -eq 'azure integrated') {
+                    if ($authType -eq 'azure integrated') {
                         # Azure AD integrated security
-                        # TODO: This is not tested
+                        # TODO: This is not tested / How can we test that?
```
Comment (wsmelton) [inline functions/Connect-DbaInstance.ps1:730]: It requires a federated Azure AD to test 😉 (which is not simple to build-out).

File: functions/Connect-DbaInstance.ps1 (hunk 1 above)
Comment (andreasjordan) [reply functions/Connect-DbaInstance.ps1:467]: But I want ALL tests, so all commands that run, to use this. I would have to put it in some global Appveyor config?

Comment (potatoqualitee) [reply functions/Connect-DbaInstance.ps1:467]: yeah, let me figure this out

Comment (potatoqualitee) [reply functions/Connect-DbaInstance.ps1]: done! added to \tests\appveyor.pester.ps1

**Comment by potatoqualitee** [comment]
Added a new containers test that performs a migration and sets up mirroring and it worked 🥳  

It performs a migration using the old code path then the new code path!

https://github.com/sqlcollaborative/dbatools/pull/7420/checks?check_run_id=2978360938

**Comment by potatoqualitee** [comment]
Something we may have to keep an eye on: https://ci.appveyor.com/project/sqlcollaborative/dbatools/builds/39854724/job/h6bmlcliith86qrm

It'll be a 60 minute timeout. Possibly Appveyor but possibly not.

**Comment by potatoqualitee** [comment]
Old tests pass, new tests pass, removed Config, bout to merge

**Comment by andreasjordan** [comment]
> Something we may have to keep an eye on: https://ci.appveyor.com/project/sqlcollaborative/dbatools/builds/39854724/job/h6bmlcliith86qrm
> 
> It'll be a 60 minute timeout. Possibly Appveyor but possibly not.

I don't know what's behind the link - "Build not found or access denied.". Maybe add a screenshot?

**Comment by potatoqualitee** [comment]
ahh, it was just a timeout on a Copy command. So we can generically keep an eye out for timeouts.

**Pull request merged**